templates/python-scrapy/src/middlewares.py

```diff
@@ -0,0 +1,71 @@
+import traceback
+
+from scrapy import Spider
+from scrapy.downloadermiddlewares.retry import RetryMiddleware
+from scrapy.http import Request, Response
+from scrapy.utils.response import response_status_message
+
+from apify.storages import RequestQueue
+
+from .utils import nested_event_loop, open_queue_with_custom_client, to_apify_request
+
+
+class ApifyRetryMiddleware(RetryMiddleware):
+    """
+    Basically the default Scrapy retry middleware enriched with Apify's Request Queue interaction.
+    """
+
+    def __init__(self, *args: list, **kwargs: dict) -> None:
+        super().__init__(*args, **kwargs)
+        try:
+            self._rq: RequestQueue = nested_event_loop.run_until_complete(open_queue_with_custom_client())
+        except BaseException:
+            traceback.print_exc()
+
+    def __del__(self):
+        nested_event_loop.stop()
+        nested_event_loop.close()
+
+    def process_response(self, request: Request, response: Response, spider: Spider) -> Request | Response:
+        """
+        Process the response and decide whether the request should be retried.
```

Comment:
I guess this `process_response` method name comes from some predefined interface we need to adhere to? If not, I'd expect it to have a better name and a different return type (if its purpose is to tell whether a request should be retried or not).

Reply:
Unfortunately, this is a mandatory structure for Scrapy's [Downloader Middlewares](https://docs.scrapy.org/en/latest/topics/downloader-middleware.html).